pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -526,6 +571,28 @@ void connectionOpened(final ClientCnx cnx) {
     }
 
     /**
+     * Clear the internal receiver queue and returns the message id of what was the 1st message in the queue that was
+     * not seen by the application
+     */
+    private MessageIdImpl clearReceiverQueue() {
+        List<Message> currentMessageQueue = new ArrayList<>(incomingMessages.size());
+        incomingMessages.drainTo(currentMessageQueue);
+        if (!currentMessageQueue.isEmpty()) {
+            MessageIdImpl nextMessageInQueue = (MessageIdImpl) currentMessageQueue.get(0).getMessageId();
+            MessageIdImpl previousMessage = new MessageIdImpl(nextMessageInQueue.getLedgerId(),
+                    nextMessageInQueue.getEntryId() - 1, nextMessageInQueue.getPartitionIndex());
+            return previousMessage;
+        } else if (lastDequeuedMessage != null) {
+            // If the queue was empty we need to restart from the message just after the last one that has been dequeued
+            // in the past
+            return lastDequeuedMessage;
```

Comment:
>Not exactly, if the queue was already empty, then we can rely on the lastDequeuedMessage, but if the queue has messages, it means the application is actually getting those messages and then updating the lastDequeuedMessage

I couldn't understand actual reason so, sorry but let's say queue received msgs 1,2,3,4. and `Reader` read msgs 1 & 2. So, as soon as reader reads msg=2, `lastDequeuedMessage` gets updated with value=2 and queue is left with 3,4. Now, reader lost connection and reconnected so, now according to logic => `incomingMessages` drains 3,4 to `currentMessageQueue` and `previousMessage=3-1=2`. So, is there any issue if reader starts with  `lastDequeuedMessage=2` or just takes `incomingMessages.poll(0, TimeUnit.MILLISECONDS)-1 and then incomingMessages.clear()`